iceberg: json schema: accept integers as numbers (doubles)

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Bug Fixes

* Iceberg/JSON Schema: Fix a bug where we would fail to parse integers when schema indicates number (floating point).

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
iceberg: json schema: accept integers as numbers (doubles)

rockwotj commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -52,13 +52,16 @@ convert_primitive(serde::json::parser& p, const primitive_type& ft) {
                 ft));
         }
-        return iceberg::long_value(p.value_int());
     case token::value_double:
```
If we get `42.0` that should be convertable to schema int type

rockwotj replied:
https://json-schema.org/understanding-json-schema/reference/numeric#integer

TBH I think we can just accept any lossy conversion and just round to the nearest int. We're not doing validation here

nvartolomei replied:
PTAL. I think this is reasonable. Rounding would be surprising.

I think we could do better in a couple of cases (i.e. parsing `9223372036854775807.0e0` (2^63-1) as integer) but seeing that other parsers (i.e rapidjson) don't care much I'm not too concerned for now. Java does it right btw.

Will probably return to it later with rework for the precision issues too.

nvartolomei committed:
iceberg: json schema: zero fractional part as int

If the value can't be represented as int we reject it. This follows java
jackson parser behavior (assuming most of the Iceberg ecosystem).

nvartolomei committed:
iceberg: json schema: add more tests

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--cb400483-0973-4412-93b1-498fddb100e2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72874">build#72874</a></summary>


| test_class                | test_method                                       | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                         |
|:--------------------------|:--------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write                                        | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72874#01997d08-50e0-48b2-b058-7ae7b7961a3f | FLAKY         | 14/21    | upstream reliability is '82.78145695364239'. current run reliability is '66.66666666666666'. drift is 16.11479 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                   |
| ClusterRateQuotaTest      | test_client_group_consume_rate_throttle_mechanism | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72874#01997d05-34c1-4c83-8344-0767550441ec | FLAKY         | 14/21    | upstream reliability is '82.3744292237443'. current run reliability is '66.66666666666666'. drift is 15.70776 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism |
| RandomNodeOperationsTest  | test_node_operations                              | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72874#01997d08-50e8-48cc-a7cd-65ea8fa00e94 | FLAKY         | 20/21    | upstream reliability is '98.15384615384616'. current run reliability is '95.23809523809523'. drift is 2.91575 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                          |
| DisablingPartitionsTest   | test_disable                                      | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72874#01997d08-50eb-4c19-900d-ed188721d1ee | FLAKY         | 14/21    | upstream reliability is '86.8421052631579'. current run reliability is '66.66666666666666'. drift is 20.17544 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                   |
| RapidTopicRecreateTest    | test_topic_rapid_recreation                       | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72874#01997d05-34c1-4c83-8344-0767550441ec | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RapidTopicRecreateTest&test_method=test_topic_rapid_recreation                     |
</details>
<!--/body-->

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27705-v25.2.x-848 remotes/upstream/v25.2.x
git cherry-pick -x 6f65b93d00 6c650aabbe b95f04d3a7
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18008462075).
